sdk/data_persistence.go

```diff
@@ -0,0 +1,294 @@
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/big"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// To reduce rpc requests for developers, save data into local storage for future reference
+type DataPersistence struct {
+	Receipts map[string]*ReceiptPersistence     `json:"receipts,omitempty"`
+	Storages map[string]*StoragePersistence     `json:"storage_slots,omitempty"`
+	Txs      map[string]*TransactionPersistence `json:"txs,omitempty"`
+}
+
+// Used for data persistence only
+type ReceiptPersistence struct {
+	TxHash       common.Hash            `json:"tx_hash,omitempty"`
+	BlockNum     *big.Int               `json:"block_num,omitempty"`
+	BlockBaseFee *big.Int               `json:"block_base_fee,omitempty"`
+	MptKeyPath   *big.Int               `json:"mpt_key_path,omitempty"`
+	Fields       []*LogFieldPersistence `json:"fields,omitempty"`
+}
+
+// Used for data persistence only
+type LogFieldPersistence struct {
+	// The contract from which the event is emitted
+	Contract common.Address `json:"contract,omitempty"`
+	// The event ID of the event to which the field belong (aka topics[0])
+	EventID common.Hash `json:"event_id,omitempty"`
+	// the log's position in the receipt
+	LogPos uint `json:"log_index,omitempty"`
+	// Whether the field is a topic (aka "indexed" as in solidity events)
+	IsTopic bool `json:"is_topic,omitempty"`
+	// The index of the field in either a log's topics or data. For example, if a
+	// field is the second topic of a log, then FieldIndex is 1; if a field is the
+	// third field in the RLP decoded data, then FieldIndex is 2.
+	FieldIndex uint `json:"field_index,omitempty"`
+	// The value of the field in event, aka the actual thing we care about, only
+	// 32-byte fixed length values are supported.
+	Value common.Hash `json:"value,omitempty"`
+}
+
+// Used for data persistence only
+type StoragePersistence struct {
+	BlockNum     *big.Int       `json:"block_num,omitempty"`
+	BlockBaseFee *big.Int       `json:"block_base_fee,omitempty"`
+	Address      common.Address `json:"address,omitempty"`
+	Slot         common.Hash    `json:"slot,omitempty"`
+	Value        common.Hash    `json:"value,omitempty"`
+}
+
+// Used for data persistence only
+type TransactionPersistence struct {
+	Hash         common.Hash `json:"hash,omitempty"`
+	BlockNum     *big.Int    `json:"block_num,omitempty"`
+	BlockBaseFee *big.Int    `json:"block_base_fee,omitempty"`
+	MptKeyPath   *big.Int    `json:"mpt_key_path,omitempty"`
+	LeafHash     common.Hash `json:"leaf_hash,omitempty"`
+}
+
+func generateReceiptKey(receipt ReceiptData, srcChainId uint64) string {
```

Comment:
This looks ambiguous. Consider using `json.Marshal()` then `sha256.Sum256` for all the keys.